File: trie.py

class TrieNode(object):
    """node of trie tree"""
    def __init__(self):
        super(TrieNode, self).__init__()
        self.ch = 0
        self.branch = 0
        self.times = 0
        self.child = {}

def init_trie():
    root = TrieNode()
    return root

def insert_trie(root, strs):
    slen = len(strs)
    if slen == 0:
        root.times += 1
        return
    i = 0
    idx = 0
    p = root
    p.times += 1
    while (i < slen):
        idx = strs[i]
        if idx in p.child:
            p = p.child[idx]
            p.times += 1
            i += 1
        else:
            tmp = TrieNode()
            tmp.ch = strs[i]
            tmp.times = 1
            p.branch += 1
            p.child[idx] = tmp
            p = tmp
            i += 1

def longest_common_prefix(contents):
    n = len(contents)
    if n == 0:
        return ''

    ret = ''
    root = init_trie()
    for x in contents:
        if len(x) < 2:
            break;
        print (x)
        insert_trie(root, x)
        i = 0
        p = root
        while i < len(contents[0]) and p.branch == 1 and p.times == n:
            p = p.child[contents[0][i]]
            i += 1
        if p.times < n:
            i -= 1
        print (contents[0][0: i])
        if i > len(ret):
            ret = contents[0][0: i]

    return ret



print (longest_common_prefix(sorted([['i', 'C', 't', 'S'], ['i', 'C', 't', 'S', 'e', 'S'], ['a']])))

File: trie.cpp
class Solution {
  public:
    typedef struct node {
        char ch;
        int branch ; //记录分支树，方便最后查询
        int times;
        node *child[26];
    } node, *Trie;

    Trie init_Trie() {
        Trie root = (node *)malloc(sizeof(node)); //Trie树的根节点不存储数据
        root->branch = 0;
        root->times = 0;

        for (int i = 0; i < 26; i++) {
            root->child[i] = NULL;
        }

        return root;
    }

    void insert_Trie(Trie root, const string str) {
        int n = str.length();

        if (n == 0) {
            root->times ++;
            return;
        }

        int i = 0;
        int idx;
        node *p = root;
        root->times++;

        while (i < n) {
            idx = str[i] - 'a';

            if (p->child[idx] != NULL) {
                p = p->child[idx];
                p->times ++;
                i++;
            } else {
                node *tmp = (node *)malloc(sizeof(node));
                tmp->ch = str[i];
                tmp->branch = 0;
                tmp->times = 1;

                for (int j = 0; j < 26; j++) {
                    tmp->child[j] = NULL;
                }

                p->branch ++;
                p->child[idx] = tmp;
                p = tmp;
                i++;
            }
        }
    }

    string longestCommonPrefix(vector<string> &strs) {
        int n = strs.size();

        if (n == 0) {
            return "";
        }

        int i;
        Trie root = init_Trie();

        for (i = 0; i < n; i++) {
            insert_Trie(root, strs[i]);
        }

        i = 0;
        node *p = root;

        while (i < strs[0].length() && p->branch == 1 && p->times == n) {
            p = p->child[strs[0][i] - 'a'];
            i++;
        }

        if (p->times < n) {
            i--;
        }

        return strs[0].substr(0, i);
    }

};
